checkmd5.py

```diff
@@ -0,0 +1,163 @@
+"""\
+Run check and update of md5sum identical files.
+
+Example.
+
+    %(prog)s --username ACCESS_KEY_ID --password SECRET_ACCESS_KEY \\
+        --out experiments_statuses.log https://www.encodeproject.org
+"""
+import datetime
+import json
+import sys
+import subprocess
+from collections import defaultdict
+from urllib.parse import urljoin
+import requests
+from slackclient import SlackClient
+
+EPILOG = __doc__
+
+def run(out, url, username, password, bot_token=None, dry_run=False):
+    session = requests.Session()
+    session.auth = (username, password)
+    session.headers['Accept'] = 'application/json'
+
+    dr = ""
+    if dry_run:
+        dr = "-- Dry Run"
+
+    version = '0.01'
+
+    try:
+        ip_output = subprocess.check_output(
+            ['hostname'], stderr=subprocess.STDOUT
+        ).strip()
+        ip = ip_output.decode(errors='replace').rstrip('\n')
+    except subprocess.CalledProcessError:
+        ip = ''
+
+    initiating_run = (
+        'STARTING matching md5sum files detection, version {} ({}) ({}): {} at {}'
+    ).format(
+        version,
+        url,
+        dr,
+        ip,
+        datetime.datetime.now()
+    )
+
+    out.write(initiating_run + '\nFile uuid\tmd5sum\tMatching md5sum files\n')
+    out.flush()
+    if bot_token:
+        sc = SlackClient(bot_token)
+        sc.api_call(
+            "chat.postMessage",
+            channel="#bot-reporting",
+            text=initiating_run,
+            as_user=True,
+        )
+
+    graph = []
+    r = session.get(
+        urljoin(
+            url,
+            '/search/?type=File&field=uuid&field=status&field=md5sum&limit=all'
+        )
+    )
+    try:
+        r.raise_for_status()
+    except requests.HTTPError:
+        return
+    else:
+        graph = r.json()['@graph']
+
+    excluded_statuses = ['uploading', 'upload failed', 'content error']
+    md5dictionary = defaultdict(set)
+    for f in graph:
+        if f.get('status') not in excluded_statuses:
+            md5 = f.get('md5sum')
+            md5dictionary[md5].add(f.get('uuid'))
```

Comment:
These two lines can potentially add an entry to `md5dictionary` with a key of `None` if the file does not have an md5sum, is that OK?

Reply:
you are right, it is not OK. Fixed.